Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift

```diff
@@ -197,9 +204,19 @@ public final class EventLoopConnectionPool<Source> where Source: ConnectionPoolS
             // connections are exhausted, we must wait for one to be returned
             logger.debug("Connection pool exhausted on this event loop, adding request to waitlist")
             let promise = eventLoop.makePromise(of: Source.Connection.self)
-            self.waiters.append((logger, promise))
+            let uuid = UUID()
+            self.waiters.append((logger, promise, uuid))
+            
+            let task = eventLoop.scheduleTask(in: self.creationTimeout) { [weak self, logger, promise] in
+                logger.error("Connection creation timed out. This might indicate a connection deadlock in your application.")
+                if let idx = self?.waiters.firstIndex(where: { $0.2 == uuid }) {
```

Comment:
Both the search and the remove would have linear runtime each. That is not ideal, but hopefully this occurs rarely enough to not be a bottleneck.

Unless in those 10 seconds before timing out so many waiters accumulate that it actually does become a bottleneck... When stuff fails, it often all fails at once, after all.